### PR TITLE
support for native amp

### LINKED ISSUE #1336: Native Amp Support
Native automatic mixed precision support (`torch.cuda.amp`) is finally merged:
https://pytorch.org/docs/master/amp.html
https://pytorch.org/docs/master/notes/amp_examples.html
Apex Amp has many known pain points (extension builds, forward/backward compatibilty, DataParallel support, flaky checkpointing, i don’t even know if it can be hacked to handle double backward/gradient penalty, others…). `torch.cuda.amp` repairs all these, the interface is more flexible and intuitive, and the tighter integration brings more future performance optimizations into scope.

If you want to talk about adding `torch.cuda.amp` to Lightning, with an eye towards it becoming the true source of mixed precision and replacing Apex, message me on Pytorch slack anytime. I pinged you there as well, but I’m not sure if you monitor it habitually.

### LINKED ISSUE #1337: Native Amp Support
Native automatic mixed precision support (`torch.cuda.amp`) is finally merged:
https://pytorch.org/docs/master/amp.html
https://pytorch.org/docs/master/notes/amp_examples.html
Apex Amp has many known pain points (extension builds, forward/backward compatibilty, DataParallel support, flaky checkpointing, i don’t even know if it can be hacked to handle double backward/gradient penalty, others…). `torch.cuda.amp` fixes all these, the interface is more flexible and intuitive, and the tighter integration brings more future performance optimizations into scope.

If you want to talk about adding `torch.cuda.amp` to Lightning, with an eye towards it becoming the true source of mixed precision and replacing Apex, message me on Pytorch slack anytime. I pinged you there as well, but I’m not sure if you monitor it habitually.

### PR DESCRIPTION
**Pull request by williamFalcon**
Fixes #1336 
Fixes #1337 

@mcarilli  mind taking a look?

## Issue 1
We have a slight issue with the DP API... 

```python
MyModel(nn.Module):
    ...
    @autocast()
    def forward(self, input):
       ...

# Alternatively
MyModel(nn.Module):
    ...
    def forward(self, input):
        with autocast():
            ...
```

@ethanwharris suggested a way around this which we have in the PR
```python
original_fwd = model.forward
model.forward = autocast()(model.forward)

# train and stuff
# ...

model.forward = original_fwd
```

## Issue 2
How do we save the state of the scaling factor to resume training?
@mcarilli

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1159,7 +1159,14 @@ def optimizer_step(self, current_epoch, batch_idx, optimizer,
         elif isinstance(optimizer, torch.optim.LBFGS):
             optimizer.step(second_order_closure)
         else:
-            optimizer.step()
+            if self.use_amp and self.use_native_amp:
+                self.trainer.scaler.step(optimizer)
```
Comment (williamFalcon) [inline pytorch_lightning/core/lightning.py:1163]: @mcarilli what happens with an optimizer that needs a second_order_closure?
@soumith

File: pytorch_lightning/trainer/distrib_parts.py [hunk 2]
```diff
@@ -517,9 +519,14 @@ def dp_train(self, model):
 
         model.cuda(self.root_gpu)
 
+        # hack forward to do autocast for the user
+        if self.use_amp and self.use_native_amp:
+            model.forward = torch.cuda.amp.autocast()(model.forward)
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/distrib_parts.py:524]: @ethanwharris put your suggestion here. see a better place this should be done?

File: pytorch_lightning/trainer/distrib_parts.py [hunk 3]
```diff
@@ -540,6 +549,9 @@ def dp_train(self, model):
 
         self.run_pretrain_routine(model)
 
+        # when training completes give back the forward
+        model.forward = model_autocast_original_forward
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/distrib_parts.py:553]: @ethanwharris

**Comment by Borda** [comment]
since which version is amp in pytorch native?

**Comment by williamFalcon** [comment]
1.6. but we don’t need to explicitly check. we can test the properties as i did

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/trainer/distrib_parts.py:526]: Not that I can see, LGTM :)

**Comment by mcarilli** [comment]
> How do we save the state of the scaling factor to resume training?

[scaler.state_dict](https://pytorch.org/docs/master/amp.html#torch.cuda.amp.GradScaler.state_dict)
[scaler.load_state_dict](https://pytorch.org/docs/master/amp.html#torch.cuda.amp.GradScaler.load_state_dict)

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -138,11 +138,20 @@ def backward(self, use_amp, loss, optimizer):
                 else:
                     loss.backward()
 
+        .. note:: with PyTorch 1.6+ + precision=16 + multiple optimizers, set .backward(retrain_graph=True)
```
Comment (mcarilli) [inline pytorch_lightning/core/hooks.py:141]: The [example](https://pytorch.org/docs/master/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers) is misleading, I guess.  The `retain_graph=True` bit has nothing to do with Amp, it's only present because both losses interleave outputs from multiple models.  Both backward passes use the same model graphs, so the first backward() must not tear them down.  `retain_graph=True` would be necessary with or without Amp.  That's unclear and maybe I should either change the example snippet so `retain_graph=True` is not needed, or add a comment clarifying that `retain_graph=True` there is not Amp-related.

File: pytorch_lightning/core/hooks.py [hunk 5]
```diff
@@ -138,11 +138,20 @@ def backward(self, use_amp, loss, optimizer):
...
         """
         if trainer.precision == 16:
-
             # .backward is not special on 16-bit with TPUs
-            if not trainer.on_tpu:
+            if trainer.on_tpu:
+                return
+
+            if self.trainer.use_native_amp:
+                # don't forget to retain graph on backward with multiple optimizers
```
Comment (mcarilli) [inline pytorch_lightning/core/hooks.py:150]: also remove this comment, see https://github.com/PyTorchLightning/pytorch-lightning/pull/1561/files#r413230111.

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (mcarilli) [reply pytorch_lightning/core/lightning.py:1163]: Closure use [isn't supported](https://pytorch.org/docs/master/amp.html#torch.cuda.amp.GradScaler.step) right now.  Making closures with dynamic gradient scaling is tricky, but not impossible.  I'm waiting for user demand before I decide if and how to implement it, so if people start asking for it, please file a Pytorch issue and assign mcarilli or add module label "amp (automated mixed precision)"

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (mcarilli) [reply pytorch_lightning/trainer/distrib_parts.py:526]: As in the [example](https://pytorch.org/docs/master/notes/amp_examples.html#dataparallel-in-a-single-process), if you intend to wrap with DataParallel, modifying model.forward to enable autocast should occur _before_ wrapping with DataParallel.  Based on code below, it looks like that's what's happening, so this is probably ok here, unless the user already wrapped the model in DataParallel before passing it in.

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -487,20 +487,18 @@ def __init__(
         self.determine_data_use_amount(train_percent_check, val_percent_check,
                                        test_percent_check, overfit_pct)
 
-        # 16 bit mixed precision training using apex
+        # AMP init
+        # These are the only lines needed after v0.8.0
+        # we wrap the user's forward with autocast and give it back at the end of fit
+        self.autocast_original_forward = None
+        self.use_native_amp = hasattr(torch.cuda, "amp") and hasattr(torch.cuda.amp, "autocast")
```
Comment (mcarilli) [inline pytorch_lightning/trainer/trainer.py:494]: this is a good way to check.

File: pytorch_lightning/core/hooks.py [hunk 7]
```diff
@@ -138,11 +138,20 @@ def backward(self, use_amp, loss, optimizer):
...
+                self.trainer.scaler.scale(loss).backward()
```
Comment (mcarilli) [inline pytorch_lightning/core/hooks.py:151]: The `enabled=True|False` arguments to `GradScaler` and `autocast` help you integrate native amp without branches, ie, instead of
```
if self.trainer.use_native_amp:
    self.trainer.scaler.scale(loss).backward()
...
else:
    loss.backward()
```
you can construct `scaler` via `GradScaler(enabled=(use_native_amp and precision = 16))` then write
```
    self.trainer.scaler.scale(loss).backward()
```
to serve all cases (`scale` becomes a no-op if `scaler` is not enabled).

This applies more generally to the rest of the integration, but it's purely for convenience and since you've already have branches written, it's a matter of preference if you want to keep them.

Comment (williamFalcon) [reply pytorch_lightning/core/hooks.py:151]: what about optimizers with closures?

Comment (mcarilli) [reply pytorch_lightning/core/hooks.py:151]: Closures aren't supported.  I haven't had anyone actually ask for one yet. https://github.com/PyTorchLightning/pytorch-lightning/pull/1561#discussion_r413237187
In principle I could implement closure handling entirely within, for example, `scaler.step(optimizer, closure=...)` but it's tricky.

Comment (williamFalcon) [reply pytorch_lightning/core/hooks.py:151]: ok. we can throw an error for amp+second order optimizers not yet supported

Comment (mcarilli) [reply pytorch_lightning/core/hooks.py:151]: as you know, LBFGS is the only native optimizer that needs a closure, but other optimizers permit a closure.  I'm not sure if you provide the option for non-LBFGS optimizers to opt-into a closure, but if yes, you should include guidance that people must prefer non-closure usage for all optimizers if Amp is enabled.

Custom optimizers will also work as long as they apply gradients in a `step()` method (and that step method doesn't need a closure).

Comment (williamFalcon) [reply pytorch_lightning/core/hooks.py:148]: got it. 
We'll keep it forked for now. Once we deprecate apex in 0.8.0 we'll do as you suggested!

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_parts.py:537]: we wrap it for the user. this keeps all the distributed stuff abstracted from their research code.

**Comment by williamFalcon** [comment]
@Borda these tests are failing bc amp is not installed... did we remove amp?

File: pytorch_lightning/trainer/training_io.py [hunk 8]
```diff
@@ -281,6 +281,10 @@ def restore(self, checkpoint_path: str, on_gpu: bool):
         if on_gpu:
             model.cuda(self.root_gpu)
 
+        # restore amp scaling
+        if self.use_amp and self.use_native_amp and 'native_amp_scaling_state' in checkpoint:
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_io.py:285]: @mcarilli sanity check this loading?

File: pytorch_lightning/trainer/training_io.py [hunk 9]
```diff
@@ -316,6 +320,10 @@ def dump_checkpoint(self):
 
         checkpoint['state_dict'] = model.state_dict()
 
+        # restore native amp scaling
+        if self.use_amp and self.use_native_amp and 'native_amp_scaling_state' in checkpoint:
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_io.py:324]: @mcarilli sanity check this saving?

Comment (mcarilli) [reply pytorch_lightning/trainer/training_io.py:324]: `state_dict` is a callable, as for modules and optimizers, so `checkpoint['native_amp_scaling_state'] = self.scaler.state_dict()` is what you want.
`checkpoint['native_amp_scaling_state'] = self.scaler.state_dict` would stash the bound-method object itself :P

Comment (mcarilli) [reply pytorch_lightning/trainer/training_io.py:324]: Also you should make sure `state_dict()` is retrieved either at the very beginning of an iteration (before any `scaler` method calls) or at the very end (after `scaler.update()`), and that the model and optimizer state dicts are saved at that same spot.

File: pytorch_lightning/trainer/training_io.py (hunk 8 above)
Comment (mcarilli) [reply pytorch_lightning/trainer/training_io.py:285]: Looks good if you fix the saving https://github.com/PyTorchLightning/pytorch-lightning/pull/1561/files#r413418705

Like saving, loading should occur either at the very beginning of an iteration (before any training-related `scaler` calls for that iteration) or the very end of an iteration, after `update()`.  It doesn't make a lot of sense to load state dicts at the end of an iteration, but if the saved state originated from a `scaler.state_dict()` call at the end of, say, iteration 1000 ("end of" meaning after iteration 1000's call to `scaler.update()`), then it's ok to call load_state_dict to resume at the beginning of iteration 1001.

File: pytorch_lightning/trainer/training_io.py [hunk 10]
```diff
@@ -316,6 +320,10 @@ def dump_checkpoint(self):
...
+            checkpoint['native_amp_scaling_state'] = self.scaler.state_dict
```
Comment (mcarilli) [inline pytorch_lightning/trainer/training_io.py:325]: `checkpoint['native_amp_scaling_state'] = self.scaler.state_dict()`

File: pytorch_lightning/trainer/training_io.py (hunk 9 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/training_io.py:324]: i thought it was a property haha. and i can’t find the docs to see if it’s a method ;)

Comment (mcarilli) [reply pytorch_lightning/trainer/training_io.py:324]: lol i see.  it's consistent with the rest

Comment (mcarilli) [reply pytorch_lightning/trainer/training_io.py:324]: Another thing to consider is that with `torch.cuda.amp`, it's permissible to
- load a checkpoint from a model + optimizer not trained with Amp, and resume training with Amp enabled, or
- load a checkpoint from a model + optimizer trained with Amp, and resume training without Amp.

I think your `if` criteria are flexible enough that that may naturally happen with the right user args but I'm not sure just from looking at it.

**Comment by Borda** [comment]
> @Borda these tests are failing bc amp is not installed... did we remove amp?

probably, unfortunately, it happened here with Horovoed https://github.com/PyTorchLightning/pytorch-lightning/pull/1529#issuecomment-618068439

File: pytorch_lightning/trainer/training_io.py (hunk 9 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/training_io.py:324]: yeah this code works.

## Case 1: Train with amp, load amp
works fine

## case 2: Train amp, load and not use amp
in this case, lightning loads the amp state but amp is disabled so user doesn't use it at all

## case 3: train regular, resume regular
works fine

## case 4: train regular, resume with amp
in this case the checkpoint has no amp state and model starts normal but on amp.

File: .drone.yml [hunk 11]
```diff
@@ -31,7 +31,7 @@ steps:
     - pip install pip -U
     - pip --version
     - nvidia-smi
-    # - bash ./tests/install_AMP.sh
+    - bash ./tests/install_AMP.sh
```
Comment (Borda) [inline .drone.yml:34]:
```suggestion
    #- bash ./tests/install_AMP.sh
```
